### PR TITLE
added property replacer to annotation definition participant

### PR DESCRIPTION
**Pull request by AlexXuChen**
Added property replacer method to annotation definition participant with static property replacer strategy class.

References redhat-developer/quarkus-ls/pull/458

Signed-off-by: Alexander Chen <alchen@redhat.com>

### CODE REVIEW

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/commons/PropertyReplacerStrategy.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.function.Function;
+
+/**
+ *
+ * Utility class for replacing property expressions
+ *
+ */
+public class PropertyReplacerStrategy {
+
+	public static Function<String, String> createNullReplacer() {
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/commons/PropertyReplacerStrategy.java:25]: A public static field should be enough to avoid creating the fuction every time. In other words, I would write:

```java
public static final Function<String, String> NULL_REPLACER = propertyKey -> propertyKey;

public static final Function<String, String> BRACKET_REPLACER = propertyKey -> propertyKey.replace("{", "").replace("}", "");

...
```

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/definition/AbstractAnnotationDefinitionParticipant.java [hunk 2]
```diff
@@ -47,18 +48,35 @@
 
 	private final String annotationName;
 
-	private final String annotationMemberName;
+	private final String[] annotationMemberNames;
+
+	protected Function<String, String> propertyReplacer;
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/definition/AbstractAnnotationDefinitionParticipant.java:53]: `private final Function<String, String> propertyReplacer;`

and defines a propertyReplacer if you need toaccess to the property replacer in the class which extends this class.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/definition/AbstractAnnotationDefinitionParticipant.java [hunk 3]
```diff
@@ -47,18 +48,35 @@
...
-	 * @param annotationMemberName the annotation member name (ex : name)
+	 *
+	 * @param annotationName        the annotation name (ex :
+	 *                              org.eclipse.microprofile.config.inject.ConfigProperty)
+	 * @param annotationMemberNames the supported annotation member names (ex :
+	 *                              name)
 	 */
-	public AbstractAnnotationDefinitionParticipant(String annotationName, String annotationMemberName) {
+	public AbstractAnnotationDefinitionParticipant(String annotationName, String[] annotationMemberNames) {
 		this.annotationName = annotationName;
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/definition/AbstractAnnotationDefinitionParticipant.java:64]: `this(annotationName, annotationMemberNames, PropertyReplacerStrategy.NULL_REPLACER);`

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/definition/AbstractAnnotationDefinitionParticipant.java (hunk 2 above)
Comment (AlexXuChen) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/definition/AbstractAnnotationDefinitionParticipant.java:53]: I would need protected for subclass access correct?

Comment (AlexXuChen) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/definition/AbstractAnnotationDefinitionParticipant.java:53]: Oh sorry, you mean defin getPropertyReplacer?

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/java/JavaFileTextDocumentService.java [hunk 4]
```diff
@@ -287,7 +287,7 @@ public void updateCodeLensSettings(MicroProfileCodeLensSettings newCodeLens) {
 								}).collect(Collectors.toList());
 						if (isDefinitionLinkSupport()) {
 							// I don't understand
-							// return Either.forRight(locations);
+							return Either.forRight(locations);
```
Comment (angelozerr) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/java/JavaFileTextDocumentService.java:290]: It's not belong to this PR.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/definition/AbstractAnnotationDefinitionParticipant.java [hunk 5]
```diff
@@ -47,18 +49,36 @@
 
 	private final String annotationName;
 
-	private final String annotationMemberName;
+	private final String[] annotationMemberNames;
+
+	private Function<String, String> propertyReplacer;
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/definition/AbstractAnnotationDefinitionParticipant.java:54]: private final

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/commons/PropertyReplacerStrategy.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import java.util.function.Function;
+
+/**
+ *
+ * Utility class for replacing property expressions
+ *
+ */
+public class PropertyReplacerStrategy {
+
+	public static final Function<String, String> NULL_REPLACER = propertyKey -> propertyKey;
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/commons/PropertyReplacerStrategy.java:25]: It should be nice to write test with those replacer. If we need in the future implement it with an another way, if we have tests we will are sure that we will not break something.

**Comment by angelozerr** [comment]
Thanks @AlexXuChen !

**Comment by AlexXuChen** [comment]
Need to fix a bug where annotations with multiple member definition hover aren't supported.
